react/features/toolbox/components/ToolboxV2.web.js

```diff
@@ -0,0 +1,408 @@
+                            : 'icon-raised-hand' }
+                        onClick = { this._onToggleRaiseHand }
+                        tooltip = { t('toolbar.raiseHand') } />
+                </div>
+                <div className = 'button-group-center'>
+                    <ToolbarButtonV2
+                        iconName = { _audioMuted && _conference
+                            ? 'icon-mic-disabled toggled'
+                            : 'icon-microphone' }
+                        onClick = { this._onToggleAudio }
+                        tooltip = { t('toolbar.mute') } />
+                    <ToolbarButtonV2
+                        iconName = 'icon-hangup'
+                        onClick = { this._onHangup }
+                        tooltip = { t('toolbar.hangup') } />
+                    <ToolbarButtonV2
+                        iconName = { _videoMuted && _conference
+                            ? 'icon-camera-disabled toggled'
+                            : 'icon-camera' }
+                        onClick = { this._onToggleVideo }
+                        tooltip = { t('toolbar.videomute') } />
+                </div>
+                <div className = 'button-group-right'>
+                    <ToolbarButtonV2
+                        iconName = 'icon-add'
+                        onClick = { this._onShowInvite }
+                        tooltip = { _addPeopleAvailable || _dialOutAvailable
+                            ? t('addPeople.title')
+                            : t('addPeople.notAvailable') } />
+                    <InfoDialogButtonV2 />
+                    <OverflowMenuButton
+                        isOpen = { this.state.showOverflowMenu }
+                        onVisibilityChange = { this._onSetOverflowVisible }>
+                        <ul className = 'overflow-menu'>
+                            <OverflowMenuItem
+                                icon = 'icon-presentation'
+                                onClick = { this._onOpenSpeakerStats }
+                                text = { t('toolbar.speakerStats') } />
+                            <OverflowMenuItem
+                                icon = 'icon-feedback'
+                                onClick = { this._onOpenFeedback }
+                                text = { t('toolbar.feedback') } />
+                            <OverflowMenuItem
+                                icon = 'icon-link'
+                                onClick = { this._onOpenKeyboardShortcuts }
+                                text = { t('toolbar.shortcuts') } />
+                        </ul>
+                    </OverflowMenuButton>
+                </div>
+            </div>
+        );
+    }
+
+    _onHangup: () => void;
+
+    /**
+     * Dispatches an action to leave the current conference.
+     *
+     * @private
+     * @returns {void}
+     */
+    _onHangup() {
+        sendAnalytics(createToolbarEvent('hangup'));
```

Comment:
Do you think it's worth the effort to put the sendAnalytics calls into a middleware, so then it gets sent everytime it has to and then there won't be need for these two-liners, like sendAnalytics(event); dispatch(event)

Reply:
I think it's worth the effort but not for the current PR. The old web toolbars plus the existing native toolbars do not leverage middleware for analytics. They all put the analytics next to the dispatch. Analytics is one of the go-to examples for middleware, though. Maybe the lack of differentiation right now between specific toolbar interactions triggering an action and other ways of triggering an action preclude an easier switch to middleware.

This is a long way for me to be an ass and say I don't want to do it for the web toolbar rewrite. I do have a task on my plate to "add analytics to toolbar clicks" which I made to remind myself to audit the analytics. I could do it then. If it can wait, I'd prefer to wait. If you sincerely believe it should be done now, I will do it now because it is honestly reasonable to at least investigate.

Reply:
No, this is clearly a separate task, I just wanted to know your opinion. But it's worth at least to have is as a task in Jira not to forget